bam/reader.go

```diff
@@ -109,90 +153,64 @@ func (br *Reader) Read() (*sam.Record, error) {
 		return nil, io.EOF
 	}
 
-	r := errReader{r: br.r}
-	bin := binaryReader{r: &r}
-
-	// Read record header data.
-	blockSize := int(bin.readInt32())
-	r.n = 0 // The blocksize field is not included in the blocksize.
-
+	var buf [4]byte
+	n, err := br.r.Read(buf[:4])
 	// br.r.Chunk() is only valid after the call the Read(), so this
 	// must come after the first read in the record.
 	tx := br.r.Begin()
 	defer func() {
 		br.lastChunk = tx.End()
 	}()
 
-	var rec sam.Record
+	if err != nil {
+		return nil, err
+	}
+	if n != 4 {
+		return nil, errors.New("bam: truncated record")
```

Comment:
I go so far as to say invalid record at this point: "bam: invalid record: short block size".